artemis-core-client/src/main/java/org/apache/activemq/artemis/core/remoting/impl/netty/NettyConnection.java

```diff
@@ -355,12 +373,24 @@ private boolean canWrite(final int requiredCapacity) {
       return canWrite;
    }
 
-   private Object getFileObject(RandomAccessFile raf, FileChannel fileChannel, long offset, int dataSize) {
-      if (channel.pipeline().get(SslHandler.class) == null) {
-         return new NonClosingDefaultFileRegion(fileChannel, offset, dataSize);
+   private Object getFileObject(FileChannel fileChannel, long offset, int dataSize) {
+      if (USE_FILE_REGION && channel.pipeline().get(SslHandler.class) == null) {
```

Comment:
`channel.pipeline().get(SslHandler.class)` could be costy: @wy96f we don't have any chance to evaluate this before?

Reply:
I didn't notice this. I'll run load generator to verify. Great work!

Reply:
It is very subtle, because the most of the cost come from checking the volatile and that will prevent further loads/stores to be moved upfront.. anyway I'm more interested into seeing how '-Dio.netty.file.region=false'  compare with master and the zero copy version :P

Reply:
Sure thing, I'm setting up the server.

Reply:
I'm still getting on wildfly other failures, but not anything related to the file replication itself (that seems fixed), nonetheless it's happening only on this PR :(

Reply:
There is some problem with -Dio.netty.file.region=false. Some page files seemingly are not transferred. At last replication failed, the log showed:
```
2019-09-24 13:38:37,508 WARN  [org.apache.activemq.artemis.core.server] AMQ222010: Critical IO Error, shutting down the server. file=NULL, message=AMQ229114: Replication synchronization process timed out after waiting 30,000 milliseconds: ActiveMQReplicationTimeooutException[errorType=REPLICATION_TIMEOUT_ERROR message=AMQ229114: Replication synchronization process timed out after waiting 30,000 milliseconds]
        at org.apache.activemq.artemis.core.replication.ReplicationManager.sendSynchronizationDone(ReplicationManager.java:687) [artemis-server-2.11.0-SNAPSHOT.jar:2.11.0-SNAPSHOT]
        at org.apache.activemq.artemis.core.persistence.impl.journal.JournalStorageManager.startReplication(JournalStorageManager.java:699) [artemis-server-2.11.0-SNAPSHOT.jar:2.11.0-SNAPSHOT]
        at org.apache.activemq.artemis.core.server.impl.SharedNothingLiveActivation$2.run(SharedNothingLiveActivation.java:178) [artemis-server-2.11.0-SNAPSHOT.jar:2.11.0-SNAPSHOT]
        at java.lang.Thread.run(Thread.java:745) [rt.jar:1.8.0_121]

2019-09-24 13:38:37,509 WARN  [org.apache.activemq.artemis.core.server] AMQ222251: Unable to start replication: ActiveMQReplicationTimeooutException[errorType=REPLICATION_TIMEOUT_ERROR message=AMQ229114: Replication synchronization process timed out after waiting 30,000 milliseconds]
        at org.apache.activemq.artemis.core.replication.ReplicationManager.sendSynchronizationDone(ReplicationManager.java:687) [artemis-server-2.11.0-SNAPSHOT.jar:2.11.0-SNAPSHOT]
        at org.apache.activemq.artemis.core.persistence.impl.journal.JournalStorageManager.startReplication(JournalStorageManager.java:699) [artemis-server-2.11.0-SNAPSHOT.jar:2.11.0-SNAPSHOT]
        at org.apache.activemq.artemis.core.server.impl.SharedNothingLiveActivation$2.run(SharedNothingLiveActivation.java:178) [artemis-server-2.11.0-SNAPSHOT.jar:2.11.0-SNAPSHOT]
        at java.lang.Thread.run(Thread.java:745) [rt.jar:1.8.0_121]
```
I am looking at it now.

Reply:
After some analysis of the logs I see I'm getting the same...
And I have verified that the files are sent at some point (the close method is getting called after being read), but I cannot see them on the other side...

Reply:
I was betting on some accident due to the Netty flow control (highwatermark value) and to some blocking call on the replication executor...

Reply:
Yes. I set breakpoints at code ```               replicator.sendSynchronizationDone(nodeID, initialReplicationSyncTimeout, ioCriticalErrorListener);```. There was nothing at ChannelOutboundBuffer but lots of `PendingWrite` in the queue of `ChunkedWriteHandler`

![chunkedinput](https://user-images.githubusercontent.com/7719761/65497248-7ea68080-deec-11e9-858b-2e154fdd04d8.png).

Reply:
So it seems that Netty is slowing sending them...can see 2 reasons for that:
- we are sending too much for real
- the receiving broker is not reading them fast enough

I cannot find other reasons, other insights?

Reply:
We are not using file regions on xnio with this pr, but chunked nio files that are correctly handled by Netty and xnio receive just ByteBufs.

Reply:
@wy96f I've just discovered that running `org.apache.activemq.artemis.tests.integration.replication.SharedNothingReplicationTest::testReplicateFromSlowLive` until failures make it fails at some point with this same error: so we have a reproducer :)

Reply:
FYI if I change `public static final int DEFAULT_WRITE_BUFFER_HIGH_WATER_MARK = 128 * 1024;` into `public static final int DEFAULT_WRITE_BUFFER_HIGH_WATER_MARK = 2 * 1024 * 1024;` the test is not failing anymore....(or at least it can fail that easily, after ~40 rounds)

Reply:
> So it seems that Netty is slowing sending them...can see 2 reasons for that:
> 
> * we are sending too much for real
> * the receiving broker is not reading them fast enough
> 
> I cannot find other reasons, other insights?

Actually broker was not sending built up data in the queue after some point until timeout failure. It seems watermark mechanism triggered some bug resulting in not resuming transfer in `ChunkedWriteHandler`.

Reply:
I see, but I've tried to raise the highwatermark value, the outbound TCP buffer and lower the size of chunks to be 32 K getting the same error...
I will try to override the back-pressure mechanism we use on NettyConnection making it to return just true...
If we will found an error on Netty side I will send a pr there instead

Reply:
I've proceed in the investigation and I see that we are failing here: https://github.com/netty/netty/blob/21b7e29ea7c211bb2b889bae3a0c6c5d9f60fb01/handler/src/main/java/io/netty/handler/stream/ChunkedWriteHandler.java#L320

And we won't get notified anymore on [ChunkedWriteHandler::channelWritabilityChanged](https://github.com/netty/netty/blob/21b7e29ea7c211bb2b889bae3a0c6c5d9f60fb01/handler/src/main/java/io/netty/handler/stream/ChunkedWriteHandler.java#L144)

Reply:
I've found the issue: was on our side. I'm going to update this PR with the change ;)

Reply:
> I've proceed in the investigation and I see that we are failing here: https://github.com/netty/netty/blob/21b7e29ea7c211bb2b889bae3a0c6c5d9f60fb01/handler/src/main/java/io/netty/handler/stream/ChunkedWriteHandler.java#L320
> 
> And we won't get notified anymore on [ChunkedWriteHandler::channelWritabilityChanged](https://github.com/netty/netty/blob/21b7e29ea7c211bb2b889bae3a0c6c5d9f60fb01/handler/src/main/java/io/netty/handler/stream/ChunkedWriteHandler.java#L144)
> 
> It means to me that:
> 
> 1. The receiver side has stopped reading data ie effectively the channel is not writeable anymore
> 2. The sender side (on Netty) has some bug while marking progress and/or propagate writeability changes

https://github.com/netty/netty/blob/21b7e29ea7c211bb2b889bae3a0c6c5d9f60fb01/handler/src/main/java/io/netty/handler/stream/ChunkedWriteHandler.java#L320 would not be called as `doFlush` is running in netty executor and channel is still not writable after writing message.
Should not be 1 bcs `netstat` showed no backlog. I found `channelWritabilityChanged` was never called during replication.

Reply:
We were missing this https://github.com/apache/activemq-artemis/pull/2845/commits/438c0fb0c69596cbd02f5ec6337c934b331aea96#diff-917ee65648802d0c63faa660eb9f8debR68
to propagate a writeable channel and resume sending

Reply:
I'm finishing running the tests, but I believe you can try to run your load generator as well now (fingers crossed!)